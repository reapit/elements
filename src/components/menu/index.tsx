export * from './menu'
export * from './menu.atoms'
export * from './styles'
