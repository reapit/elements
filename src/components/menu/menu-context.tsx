import { createContext, useContext, type FC } from 'react'
import { useMenu } from '../../hooks/use-menu'

const MenuContext = createContext<useMenu | null>(null)

const MenuProvider: FC = ({ children }) => {
  const menu = useMenu()

  return <MenuContext.Provider value={menu}>{children}</MenuContext.Provider>
}

export { MenuContext, MenuProvider }

export const useMenuContext = () => {
  const context = useContext(MenuContext)
  if (!context) {
    throw new Error('useMenuContext must be used within a MenuProvider')
  }
  return context
}
