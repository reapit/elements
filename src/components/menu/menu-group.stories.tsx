import type { Meta, StoryObj } from '@storybook/react'
import {
  ElMenuItemButton,
  ElMenuItemContainer,
  ElMenuItemGroup,
  ElMenuItemGroupTitle,
  ElMenuItemLink,
  MenuItem,
  MenuItemGroup,
} from '.'
import { TextSM } from '../typography'

const meta: Meta<typeof MenuItemGroup> = {
  title: 'Components/Menu/Menu Item Group',
  component: MenuItemGroup,
  argTypes: {
    title: {
      type: 'string',
    },
  },
}

export default meta
type Story = StoryObj<typeof MenuItemGroup>

export const StylesOnlyUsage: StoryObj = {
  render: () => {
    return (
      <ElMenuItemGroup>
        <li>
          <ElMenuItemGroupTitle>Group Title</ElMenuItemGroupTitle>
        </li>
        <ElMenuItemContainer>
          <ElMenuItemLink href="/#">
            <TextSM hasNoMargin>This is a Link</TextSM>
          </ElMenuItemLink>
        </ElMenuItemContainer>
        <ElMenuItemContainer>
          <ElMenuItemLink href="/#">
            <TextSM hasNoMargin>This is a Link</TextSM>
          </ElMenuItemLink>
        </ElMenuItemContainer>
        <ElMenuItemContainer>
          <ElMenuItemButton onClick={() => 0}>
            <TextSM hasNoMargin>This is a Button</TextSM>
          </ElMenuItemButton>
        </ElMenuItemContainer>
      </ElMenuItemGroup>
    )
  },
}

export const ReactVersion: Story = {
  args: {
    title: 'Group Title',
  },
  render: ({ title }) => (
    <MenuItemGroup title={title}>
      <MenuItem href="#">This is a Link</MenuItem>
      <MenuItem onClick={() => 0}>This is a Button</MenuItem>
    </MenuItemGroup>
  ),
}
