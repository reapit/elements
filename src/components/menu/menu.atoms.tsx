import type {
  AnchorHTMLAttributes,
  ButtonHTMLAttributes,
  FC,
  HTMLAttributes,
  MouseEventHandler,
  ReactNode,
} from 'react'
import { ElMenuItemAnchor, ElMenuItemButton, ElMenuItemGroup, ElMenuItemGroupTitle, ElMenuList } from './styles'

interface CommonMenuItemProps {
  children?: ReactNode
  /**
   * Whether the menu is closed when clicking this item
   *
   * @example
   * <MenuItem closeMenu={false}>This won't close the menu</MenuItem>
   *
   * @default true
   */
  closeMenu?: boolean
}

interface MenuItemAsButtonElementProps extends CommonMenuItemProps, ButtonHTMLAttributes<HTMLButtonElement> {
  href?: never
  disabled?: boolean
  onClick?: MouseEventHandler<HTMLButtonElement>
}

interface MenuItemAsAnchorElementProps extends CommonMenuItemProps, AnchorHTMLAttributes<HTMLAnchorElement> {
  /** MenuItemAsAnchor currently doesn't support disabled state */
  disabled?: never
}

export type MenuItemProps = MenuItemAsButtonElementProps | MenuItemAsAnchorElementProps

/**
 * The `MenuItemGroup` component is a wrapper for `MenuItem` which has optional label
 * can be used for grouping multiple type of list item e.g RadioItem (in future).
 */
export const MenuItemGroup: FC<
  HTMLAttributes<HTMLDivElement> & {
    label?: string
  }
> = ({ children, label, ...rest }) => {
  return (
    <ElMenuItemGroup role="group" {...rest}>
      {!!label && <ElMenuItemGroupTitle>{label}</ElMenuItemGroupTitle>}
      {children}
    </ElMenuItemGroup>
  )
}

export const MenuItem: FC<MenuItemProps> = ({ children, disabled, closeMenu = true, ...rest }) => {
  if (!isItemAsButtonElement(rest)) {
    return (
      <ElMenuItemAnchor role="menuitem" data-close-menu={closeMenu} {...(rest as MenuItemAsAnchorElementProps)}>
        {children}
      </ElMenuItemAnchor>
    )
  }

  return (
    <ElMenuItemButton role="menuitem" data-close-menu={closeMenu} aria-disabled={disabled} {...rest}>
      {children}
    </ElMenuItemButton>
  )
}

export const MenuList: FC<HTMLAttributes<HTMLDivElement>> = ({ children, ...rest }) => (
  <ElMenuList role="menu" {...rest}>
    {children}
  </ElMenuList>
)

function isItemAsButtonElement(props: MenuItemProps): props is MenuItemAsButtonElementProps {
  return !props.href
}
