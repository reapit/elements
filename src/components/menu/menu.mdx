import { Canvas, Meta, Controls } from '@storybook/blocks'
import * as MenuStories from './menu.stories'
import { RenderHtmlMarkup } from '../../storybook/render-html-markup'

<Meta of={MenuStories} />

# Menu

The Menu component is a panel used to display a list of buttons and/or anchors which can be ordered in Menu Groups

Each Menu Group can have a title shown above the list

More features will be added to Menu in the future

<Controls />

## Default

By default, the `Menu` is left-aligned, but this can be adjusted using the `data-alignment`. Additionally (for React user), the vertical position of the Menu is automatically adjusted based on the available space below the trigger and `window` each time the menu is opened (can use `yOffset` to have more adjustment).

`closeMenu`={false} can be set to keep the menu open when it is clicked.

Note: `closeMenu` and `yOffset` are react only props.

<Canvas of={MenuStories.Default} />

<RenderHtmlMarkup of={MenuStories.Default} />

## With Alignment

Example for `Menu` with right Alignment.

<Canvas of={MenuStories.WithCustomAlignment} />

<RenderHtmlMarkup of={MenuStories.WithCustomAlignment} />
