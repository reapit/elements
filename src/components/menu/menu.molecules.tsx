import type { FC } from 'react'
import { MenuItemContainer, type MenuItemContainerProps } from './menu.atoms'
import { ElMenuItemContent, ElMenuItemLabel } from './styles'

interface MenuItemProps extends Omit<MenuItemContainerProps, 'children'> {
  label?: string
}

export const MenuItem: FC<MenuItemProps> = ({ label, ...props }) => {
  return (
    <MenuItemContainer {...(props as MenuItemContainerProps)}>
      <ElMenuItemContent>
        <ElMenuItemLabel>{label}</ElMenuItemLabel>
      </ElMenuItemContent>
    </MenuItemContainer>
  )
}
