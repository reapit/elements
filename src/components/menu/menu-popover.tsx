import { FC, MutableRefObject, useRef } from 'react'
import { useClickOutside } from '../../hooks/use-click-outside'
import { useMenuContext } from './menu-context'
import { ElMenuPopover } from './styles'

export const MenuPopover: FC = ({ children }) => {
  const { isOpen, closeMenu, getPopoverProps } = useMenuContext()
  const popoverRef = useRef<HTMLDivElement>(null)

  useClickOutside(popoverRef, closeMenu)

  if (isOpen) {
    const onClick = (event) => {
      // close menu if clicked target has parent with below selector
      const menuItem = event.target.closest('[data-close-menu="true"]')
      if (menuItem) closeMenu()
    }

    return (
      <ElMenuPopover ref={popoverRef as MutableRefObject<HTMLDivElement>} {...getPopoverProps()} onClick={onClick}>
        {children}
      </ElMenuPopover>
    )
  }

  return null
}

export const MenuTrigger = ({ children }) => {
  const { getTriggerProps, isOpen } = useMenuContext()

  return children({ getTriggerProps, isOpen })
}
