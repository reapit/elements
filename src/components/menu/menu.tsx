import React, { type HTMLAttributes } from 'react'
import { MenuProvider } from './menu-context'
import { MenuPopover, MenuTrigger } from './menu-popover'
import { MenuItem, MenuItemGroup, MenuList } from './menu.atoms'
import { ElMenu } from './styles'

type MenuFC = React.FC<HTMLAttributes<HTMLDivElement> & { 'data-alignment'?: 'left' | 'right' }> & {
  Item: typeof MenuItem
  Group: typeof MenuItemGroup
  Trigger: typeof MenuTrigger
  Popover: typeof MenuPopover
  List: typeof MenuList
}

const Menu: MenuFC = ({ children, ...props }) => {
  return (
    <MenuProvider>
      <ElMenu {...props}>{children}</ElMenu>
    </MenuProvider>
  )
}
Menu.Item = MenuItem
Menu.Group = MenuItemGroup
Menu.Trigger = MenuTrigger
Menu.Popover = MenuPopover
Menu.List = MenuList

export { Menu }
