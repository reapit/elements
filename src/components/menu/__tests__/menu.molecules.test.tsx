import { fireEvent, render, screen } from '@testing-library/react'
import { handlePopoverPosition, Menu } from '../menu.molecules'
import { useRef } from 'react'

describe('Menu component', () => {
  it('should render Menu and match snapshot', () => {
    const { asFragment } = render(
      <Menu>
        <Menu.List>
          <Menu.Group title="Radio Title">
            <Menu.RadioItem checked>Item 1</Menu.RadioItem>
            <Menu.RadioItem>Item 1</Menu.RadioItem>
          </Menu.Group>
          <Menu.Group title="Group Title">
            <Menu.Item>Item 1</Menu.Item>
            <Menu.Item>Item 2</Menu.Item>
          </Menu.Group>
        </Menu.List>
      </Menu>,
    )
    expect(asFragment()).toMatchSnapshot()
  })
})

describe('Menu Trigger component', () => {
  it('should render Trigger and match snapshot', () => {
    const { asFragment } = render(
      <Menu>
        <Menu.Trigger>{({ getTriggerProps }) => <button {...getTriggerProps()}>Trigger</button>}</Menu.Trigger>
      </Menu>,
    )
    expect(asFragment()).toMatchSnapshot()
  })
})

describe('Menu Popover component', () => {
  const MockMenuPopoverComponent = () => {
    const ref = useRef(null)
    return (
      <div ref={ref}>
        <Menu>
          <Menu.Trigger>{({ getTriggerProps }) => <button {...getTriggerProps()}>Trigger</button>}</Menu.Trigger>
          <Menu.Popover containerRef={ref}>
            <div>Popover Content</div>
            <button type="button">Close</button>
          </Menu.Popover>
        </Menu>
      </div>
    )
  }

  it('should render Popover when open and match snapshot', () => {
    const { asFragment } = render(<MockMenuPopoverComponent />)

    fireEvent.click(screen.getByText('Trigger'))
    expect(asFragment()).toMatchSnapshot()
  })

  it('should close the Popover when button inside popover clicked and match snapshot', () => {
    const { asFragment } = render(<MockMenuPopoverComponent />)

    fireEvent.click(screen.getByText('Trigger'))
    fireEvent.click(screen.getByText('Close'))
    expect(asFragment()).toMatchSnapshot()
  })
})

describe('handlePopoverPosition', () => {
  let ref
  let getTriggerProps
  let getPopoverProps
  let setPopoverStyle
  let isOpen

  beforeEach(() => {
    ref = {
      current: document.createElement('div'),
    }

    const triggerBtn = document.createElement('button')
    triggerBtn.className = 'trigger-btn'
    ref.current.appendChild(triggerBtn)

    const popover = document.createElement('div')
    popover.className = 'popover'
    ref.current.appendChild(popover)

    getTriggerProps = () => ({ className: 'trigger-btn' })
    getPopoverProps = () => ({ className: 'popover' })
    setPopoverStyle = jest.fn()
    isOpen = true
  })

  it('sets popover position below the trigger button if space is available', () => {
    const triggerBtn = ref.current.querySelector('.trigger-btn')
    const popover = ref.current.querySelector('.popover')

    jest.spyOn(triggerBtn, 'getBoundingClientRect').mockReturnValue({
      top: 100,
      bottom: 150,
      height: 50,
      width: 100,
    } as DOMRect)

    jest.spyOn(popover, 'getBoundingClientRect').mockReturnValue({
      height: 200,
      width: 100,
    } as DOMRect)

    window.innerHeight = 600

    const repositionPopover = handlePopoverPosition(ref, isOpen, getTriggerProps, getPopoverProps, setPopoverStyle)
    repositionPopover()

    expect(setPopoverStyle).toHaveBeenCalledWith({ top: 53 }) // Button height (50) + gap (3)
  })

  it('sets popover position above the trigger button if there is not enough space below', () => {
    const triggerBtn = ref.current.querySelector('.trigger-btn')
    const popover = ref.current.querySelector('.popover')

    jest.spyOn(triggerBtn, 'getBoundingClientRect').mockReturnValue({
      top: 100,
      bottom: 150,
      height: 50,
      width: 100,
    } as DOMRect)

    jest.spyOn(popover, 'getBoundingClientRect').mockReturnValue({
      height: 500,
      width: 100,
    } as DOMRect)

    window.innerHeight = 400

    const repositionPopover = handlePopoverPosition(ref, isOpen, getTriggerProps, getPopoverProps, setPopoverStyle)
    repositionPopover()

    expect(setPopoverStyle).toHaveBeenCalledWith({ top: -503 }) // -Popover height (500) - gap (3)
  })

  it('does nothing if isOpen is false', () => {
    isOpen = false

    const repositionPopover = handlePopoverPosition(ref, isOpen, getTriggerProps, getPopoverProps, setPopoverStyle)
    repositionPopover()

    expect(setPopoverStyle).not.toHaveBeenCalled()
  })

  it('does nothing if ref.current is null', () => {
    ref.current = null

    const repositionPopover = handlePopoverPosition(ref, isOpen, getTriggerProps, getPopoverProps, setPopoverStyle)
    repositionPopover()

    expect(setPopoverStyle).not.toHaveBeenCalled()
  })
})
