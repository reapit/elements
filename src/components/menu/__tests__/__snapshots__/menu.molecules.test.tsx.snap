// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`MenuItem component > should render with complete features enabled and match snapshot 1`] = `
<DocumentFragment>
  <button
    class="mocked-styled-8 el-menu-item-button"
    data-close-menu="false"
    role="menuitem"
    tabindex="0"
  >
    <div
      class="el-menu-item-left-icon el-menu-item-icon"
    >
      <span>
        Left Icon
      </span>
    </div>
    <div
      class="mocked-styled-6 el-menu-item-content"
    >
      <span
        class="mocked-styled-3 el-menu-item-label-container"
      >
        <span
          class="mocked-styled-4 el-menu-item-label"
        >
          Test Menu Item
        </span>
        <span>
          badge
        </span>
      </span>
      <span
        class="mocked-styled-2 el-menu-item-supplementary-info"
      >
        Test Supplementary Info
      </span>
    </div>
    <div
      class="mocked-styled-7 el-menu-item-icon"
    >
      <span>
        Right Icon
      </span>
    </div>
  </button>
</DocumentFragment>
`;

exports[`MenuItem component > should render with required props and match snapshot 1`] = `
<DocumentFragment>
  <button
    class="mocked-styled-8 el-menu-item-button"
    data-close-menu="true"
    role="menuitem"
    tabindex="0"
  >
    <div
      class="mocked-styled-6 el-menu-item-content"
    >
      <span
        class="mocked-styled-3 el-menu-item-label-container"
      >
        <span
          class="mocked-styled-4 el-menu-item-label"
        >
          Test Menu Item
        </span>
      </span>
    </div>
  </button>
</DocumentFragment>
`;
